### PR TITLE
Semantic segmentation - Data upload

### PR DESCRIPTION
**Pull request by nbardy**
This PR modifies the way the semantic segmentation upload works. 

It changes the API from a list of masks to a dictionary of masks, so that we have a name for each mask.

It also adds the `class_labels` upload via config.

### CODE REVIEW

File: wandb/wandb_run.py [hunk 1]
```diff
@@ -661,6 +661,41 @@ def _add_viz(self, key, viz_id):
         }
         self.config.persist()
 
+    # Stores a singleton item to wandb config.
+    #
+    # A singleton in this context is a piece of data that is continually
+    # logged with the same value in each history step, but representing
```
Comment (shawnlewis) [inline wandb/wandb_run.py]: represented

File: wandb/wandb_run.py [hunk 2]
```diff
@@ -661,6 +661,41 @@ def _add_viz(self, key, viz_id):
...
+    # as a single item in the config.
+    #
+    # We do this to avoid filling up history with a lot of repeated uneccessary data
```
Comment (shawnlewis) [inline wandb/wandb_run.py]: It doesn't affect the history.

File: wandb/wandb_run.py [hunk 3]
```diff
@@ -661,6 +661,41 @@ def _add_viz(self, key, viz_id):
...
+    def _add_singleton(self, type, key, value):
+        # Wrap te value with information
+        value_extra = {
+            'type': type,
+            'key': key,
+            'value': value
+        }
+
+        if not 'singletons' in self.config['_wandb']:
+            self.config._set_wandb('singletons', {})
```
Comment (shawnlewis) [inline wandb/wandb_run.py]: Do we need this special singletons area? We could store key directly under _wandb, and then check if the value is equal with a deep compare.

File: wandb/wandb_run.py (hunk 2 above)
Comment (nbardy) [reply wandb/wandb_run.py]: That's exactly the point we're storing it here to keep it out of history. Is the wording here confusing?

File: wandb/wandb_run.py (hunk 3 above)
Comment (nbardy) [reply wandb/wandb_run.py]: No. I removed it.

File: wandb/wandb_run.py (hunk 2 above)
Comment (shawnlewis) [reply wandb/wandb_run.py]: Oh I get it.

File: wandb/wandb_run.py [hunk 4]
```diff
@@ -661,6 +661,36 @@ def _add_viz(self, key, viz_id):
...
+
+        if type in self.config['_wandb'][key]:
+            old_value = self.config['_wandb'][key][type]
+        else:
+            old_value = None
+
+        if value_extra != old_value:
+            self.config['_wandb'][key][type] = value_extra
+
+        self.config.persist()
```
Comment (shawnlewis) [inline wandb/wandb_run.py]: Shouldn't this be under the check above?

A deep comparison would probably be better for the check above.

**Comment by nbardy** [comment]
@vanpelt I'd like you to take another look at this.

File: wandb/data_types.py [hunk 5]
```diff
@@ -905,7 +908,8 @@ def to_json(self, run):
         if self._boxes:
             json_dict['boxes'] = self._boxes.to_json(run)
         if self._masks:
-            json_dict['masks'] = [mask.to_json(run) for mask in self._masks]
+            json_dict['masks'] = {
+                    k: mask.to_json(run) for (k, mask) in self._masks.items()}
```
Comment (vanpelt) [inline wandb/data_types.py:912]: I still think we should store the mask as image data.  I believe @shawnlewis agreed on the image approach.  I guess I could be swayed but it just feels really inefficient to store all the mask data in json, there will be no compression at all...

Comment (nbardy) [reply wandb/data_types.py:912]: The mask is using image data.

The `to_json` call creates a file for the mask data:

https://github.com/wandb/client/pull/883/files

**Pull request merged**